equinox/nn/_kv_cache.py

```diff
@@ -0,0 +1,73 @@
+from collections.abc import Callable
+
+import jax.lax as lax
+import jax.numpy as jnp
+from jaxtyping import Array, Float, Int
+
+from .._misc import default_int_dtype
+from .._module import field, Module
+from ._stateful import State, StateIndex
+
+
+KVCacheCallable = Callable[
+    [
+        Float[Array, "seq_length num_heads qk_size"],
+        Float[Array, "seq_length num_heads vo_size"],
+        State,
+    ],
+    tuple[
+        Float[Array, "state_length num_heads qk_size"],
+        Float[Array, "state_length num_heads vo_size"],
+        Int[Array, ""],
+        State,
+    ],
+]
+
+
+class StandardKVCache(Module):
+    key_shape: tuple[int, int, int] = field(static=True)
+    value_shape: tuple[int, int, int] = field(static=True)
+
+    autoregressive_index: StateIndex
+
+    def __init__(
+        self,
+        state_length: int,
+        num_heads: int,
+        key_size: int,
+        value_size: int,
```

Comment:
Oh, we should probably have a `dtype` argument here, since that seems to be a thing we're doing going forwards.

Reply:
Ah, right.